kep/217-resource-limit-aware-scorcing/README.md

```diff
@@ -0,0 +1,148 @@
+- node2: total request/allocatable = 6/8, total limit/allocatable = 9/8
+
+node1’s total resource limit remains the same and node2 has a total resource limit of 9. node2 will have a lower risk of resource contention with pod5 running on it. 
+
+## Terms
+
+## Proposal
+
+A custom scheduler scoring plugin scores a node by considering existing pods’ resource limits to the node's allocatable ratio on the node. It attempts to “evenly” spread or distribute pod resource limits across nodes in a cluster. 
+
+## Design Details
+
+When choosing a node to place a pod, the custom scoring plugin prefers the node that has the smallest total resource limit ratio on a node. The scoring function favors a node with the smallest over-subscription ratio.
+
+   `node limit ratio = sum of existing pods' limits on node / node’s allocatable resource`
+   
+1. Calculate each node’s raw score. A node's raw score will be negative for an over-subscription node.
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
+    
+    - For multiple resources, the raw score is the weighted sum of all allocatable resources. The resources and their weights are configurable as the plugin arguments. Like the other `NodeResources` scoring plugins, the plugin supports standard resource types: CPU, Memory, Ephemeral-Storage plus any Scalar-Resources. 
+
+2. Normalized the raw score to `[MinNodeScore , MaxNodeScore]` (e.g., [0, 100]) after all nodes are scored.
+
+    - `NormalizedScore = MinNodeScore+(RawScore-LowestRawScore)/(HighestRawScore-LowestRawScore) *(MaxNodScore - MinNodeScore)`
+
+3. Choose the node with the highest score.
+
+## Implementation
+
+- A node's allocatable can be directly obtained from `NodeInfo.Allocatable`. 
+
+- A node's resource limit of all running pods on the node is calculated from `NodeInfo.Pods`.
+
+```
+// calculateAllocatedLimit returns the allocated resource limits on a node
+ func calculateAllocatedLimit(node *framework.NodeInfo, resource v1.ResourceName) int64 {
+ 	var limit int64
+ 	for _, pod := range node.Pods {
+ 		for _, container := range pod.Pod.Spec.Containers {
+ 			limit += getLimitForResource(resource, &container.Resources.Limits)
+ 		}
+ 	}
+ 	return limit
+ }
+
+  // GetLimitForResource returns a specified resource limit for a container
+ func getLimitForResource(resource v1.ResourceName, limits *v1.ResourceList) int64 {
+ 	switch resource {
+ 	case v1.ResourceCPU:
+ 		return limits.Cpu().MilliValue()
+ 	case v1.ResourceMemory:
+ 		return limits.Memory().Value()
+ 	default:
+ 		if v1helper.IsScalarResourceName(resource) {
+ 			quantity, found := (*limits)[resource]
+ 			if !found {
+ 				return 0
+ 			}
+ 			return quantity.Value()
+ 		}
+ 	}
+ 	return 0
+ }
+```
+- For a best effort pod without specifying resource limits, a default value (e.g., the node's allocatable or configurable value) is used. 
```

Comment:
Good point. I think we can skip guaranteed pods.

Reply:
> If all pods are guaranteed, it will be equivalent to a weight of 2, but I guess it logically makes more sense by including guaranteed pods.

SG. Please add some statements to explain how Guaranteed pods will be scored.

Reply:
done.